jaxley/channels/lif.py

```diff
@@ -0,0 +1,127 @@
+# This file is part of Jaxley, a differentiable neuroscience simulator. Jaxley is
+# licensed under the Apache License Version 2.0, see <https://www.apache.org/licenses/>
+
+from typing import Dict, Optional
+from warnings import warn
+
+import jax.numpy as jnp
+
+from jaxley.channels import Channel
+from jaxley.solver_gate import save_exp
+
+
+class LIF(Channel):
+    def __init__(self, name: Optional[str] = None):
+        self.current_is_in_mA_per_cm2 = True
+        super().__init__(name)
+        self.channel_params = {
+            f"{self.name}_gLeak": 0.0003,
+            f"{self.name}_eLeak": -55,
+            f"{self.name}_vth": -50,
+            f"{self.name}_vreset": -55,
+            f"{self.name}_Tref": 2.0,  # refractory period in ms
+        }
+        self.channel_states = {
+            f"{self.name}_rem_tref": 0.0,  # time remaining in refractory period
+        }
+        self.current_name = f"{self.name}_i"
+
+        warn("This channel is not differentiable. Consider using SmoothLIF instead.")
+
+    def update_states(self, states, dt, v, params):
+        prefix = self._name
+        # Decrease refractory time
+        rem_tref = states[f"{prefix}_rem_tref"]
+        # Reset refractory timer when spike occurs
+        new_rem_tref = jnp.where(
+            v >= params[f"{prefix}_vth"],
+            params[f"{prefix}_Tref"],  # Reset to full refractory period
+            jnp.maximum(0.0, rem_tref - dt),  # Otherwise decrease by dt
+        )
+        return {f"{prefix}_rem_tref": new_rem_tref}
+
+    def compute_current(self, states, v, params):
```

Comment:
all currents and update_states should be written to handle _a single compartment_. This means that `v` will be a scalar.

Use an [select](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.select.html#jax.lax.select)-case instead. Internally, we then `vmap` the `update_states`, and `JAX` is smart enough to automatically case the `cond` to `where`.